### ISSUE TITLE
golang 1.18 generics

### ISSUE THREAD

**Issue by cchandel** [gravestench/akara#9]
Hi,
  Are you looking at refactoring to take advantage of generics in go 1.18?
Thanks

**Comment by gravestench** [comment]
_fuck yes_, I am.

**Comment by gravestench** [comment]
Actually, I'll leave this open specifically for that effort.

**Comment by amj** [comment]
very interested in this!

**Comment by gravestench** [comment]
I'm still planning on continuing work with akara and director, but I've been busy with my new job and new gaming PC 👹

I'll hopefully continue work again soon!

**Comment by amj** [comment]
@gravestench thanks!  Do you have some design ideas for how you'd want to use generics?  I think i have some ideas how they might make a typable BaseSystem but i'm betting you've already thought through where you'd put them :)

**Comment by amj** [comment]
lol, thanks!  I'm still pretty new to Go and the ecs pattern generally, so i probably have more questions than answers.
The component registration is still pretty opaque to me, e.g. in the example Movement system, `(m *MovementSystem) _Init` calls `m.RegisterComponent` which IIUC falls through to the embedded `*World` but it seems to implicitly create the ComponentFactory... which is a shame b/c as i understand it, the CF is basically the datastore for those components, which might need a different backing structure than a `map[EID]Component`, right?

So far a generic version of ComponentFactory seems good; this way i can have e.g. a generic CF backed by a `map`, one backed by a quadtree, a many-to-one CF for components shared by entities, etc.

While i've got you here, do you have any tips for how you integrated this with ebiten?  Did you run the ebiten gameloop and have it call tick() on all the systems?  or did you have e.g. an input system, render system etc and basically not use the ebiten loop?

(Also what's with subscriptions instead of channels? :)  I am new to golang but it seemed like it might make sense for systems to communicate via channels... as you can tell i have lots of questions but no opinions yet)

**Comment by gravestench** [comment]
> ... which might need a different backing structure than a map[EID]Component, right?

It certainly is an interesting idea, I have not thought about this. I'll admit, I did design this fully intending entities to have single instances of a component and no more. This decision has definitely forced me into some corners when trying to implement some things. It's also made me extremely fucking nervous about the performance implications when you have tens of millions of entities, each with large sets of components.

```golang
type ComponentFactoryStrategy int

const (
    MapStrategy ComponentFactoryStrategy = iota // default
    QuadTreeStrategy
    ManyToOneStrategy
)
```

Registration could take an optional `ComponentFactoryStrategy`, which defaults to the map strategy.

______________________

> While i've got you here, do you have any tips for how you integrated this with ebiten?

At the time of writing, Ebiten did not give you control over the main loop, you needed to give a function it would call as the game loop. This led to some ~~poor~~ magic implementations of my rendering system.

For a more complete implementation, please refer to my other project [Director](https://github.com/gravestench/director), which is a scene abstraction layer build on top of Akara. I have since switched over to `Raylib` from `Ebiten` for performance reasons, but I wouldn't mind adding the option to select which renderer/audio backing engine at init.

You can run the examples from your shell like this:
```bash
go run github.com/gravestench/director/cmd/examples/labels/scene_moving_labels@latest
```
```bash
go run github.com/gravestench/director/cmd/examples/audio/tones@latest
```
```bash
go run github.com/gravestench/director/cmd/examples/lua/shapes_from_lua@latest
```
```bash
go run github.com/gravestench/director/cmd/examples/images/from_web@latest
```
```bash
go run github.com/gravestench/director/cmd/examples/input/mouse_hitbox@latest
```

The rendering system in Director definitely needs work. Basically, scenes always have at least one camera and a viewport, the viewport is stretched across the window (but can be transformed and translated just like any other object), everything is rendered from the camera's perspective. Scenes also have a Lua state machine, and they can be created from Lua scripts.

> Also what's with subscriptions instead of channels?
If you see how to make subscriptions better, please do. I don't have much experience with channels, it was hard enough making systems operate through goroutines (thanks again @ianling !!)

Feel free to message me on Discord:
# Gravestench#4288

**Comment by amj** [comment]
Right i'm trying to get my head around a couple constraints:

1. Systems are the thing that need to do work based on their set of components.  They get access to their components via a two step: they keep a subscription to track the EIDs they need, and then they pull the component data they need out of the datastore (in e.g. `MovementSystem`, that's an embedded handle to the `componentFactory`)
2. For the world to create/provide subscriptions, it needs to also loop through all the ComponentFactories' datastores to set up initial membership.

So the component datastores need to a) provide a common interface for random subset access and still b) allow for optimized access patterns for the Systems that use them.

I'm not worried (yet) about millions of entities/components, i'm more worried about reaching 1k/10k for scenes that have lots of interactions, running n^2 distance calculations between components, spatially searching & pathfinding, etc.

E.g. if there's a game with 10k sprites w/ `Positions` and a System that updates them all by looking in each object's local neighborhood for other objects and doing computations, how does it do that efficiently?  Does the system copy all the positions into a more efficient data structure each tick?  Do i write a fancy dynamic Subscription?


Anyway coming back to the thread topic, the first part seems to be a good fit for the generics at least :)  Having duplicate `FooFactory` types for each type of component seems ripe.

I'm not sure about a FactoryStrategy pattern.  What about `RegisterComponentAndFactory(c Component, cf ComponentFactory)` on a `World` receiver?  The initialization process in the examples puts initializing the components/factories in the system initialization, but that seems weird b/c the systems don't own the datastores.

**Comment by gravestench** [comment]
> Systems are the thing that need to do work based on their set of components

They can do work without them, too! I think I have had a `Bootstrap` system in the past that init's all other systems and then disables itself, but that's generally not the norm.

__________

> For the world to create/provide subscriptions, it needs to also loop through all the ComponentFactories' datastores to set up initial membership.

The world provides subscriptions by maintaining a bit mask for each entity of which components it has. as components are added and removed for a given entity, this bit mask is maintained (this is why we need to register components, we need to allocate an index for the component type!). A subscription is just an easy way of filtering all entities in the world based on the components. As an example, a `MovementSystem` is just interested in things with a position and a velocity and would have a subscription to all entities that have both of those components.

________

> Anyway coming back to the thread topic, the first part seems to be a good fit for the generics at least :) Having duplicate FooFactory types for each type of component seems ripe.

yeah, my main goal is to just return a "concrete" component factory for the component when you register it. This prevents you from having to make a wrapper for the methods of the generic component factory which just yields the component interface for calls to `Add` and `Get`.

_________

> I'm not sure about a FactoryStrategy pattern ...

I haven't thought about this much. Take into account the way that I mentioned how subscriptions are maintained with a bit mask. how would this work when a single entity can have multiple components of the same type?

**Comment by amj** [comment]
> A subscription is just an easy way of filtering all entities in the world based on the components.

Right, got it.  And the world creates/maintains subscriptions by iterating over all the datastores.  `AddSubscription` loops through all the keys in the factories' `instances` map, `updateComponentFlags` makes the bitset by querying all the factories to see what EIDs they know about.

It was not immediately apparent to me that the bitset is derived & must be kept in sync -- the datastores represent the source of truth about what components are attached to what entity.  This is fine, i think?  The World obj is dependent on the datastores' interface, which has a bearing on whatever way you decide to do generics.  (I think it also means there's an opportunity to provide a `(w World) AddTo(EID, Component)` that updates the bitset and subscriptions without looping thru all factories?)

> how would this work when a single entity can have multiple components of the same type?

I'm proposing the opposite; many entities sharing the same component (& thus the same behavior).  Having an entity with multiple components is easy enough AIUI -- put a slice or whatever on the component struct right?

**Comment by gravestench** [comment]
> the bitset is derived & must be kept in sync

unfortunately, yes.

_______

> also means there's an opportunity to provide a `(w World) AddTo(EID, Component)`

I'm not sure I follow. Are you suggesting this in order to reduce some of the overhead incurred during per-entity component map management? I'm down with whatever optimizations you can think of!

________

> I'm proposing the opposite...

gotcha. I ended up implementing this kind feature when using akara by referencing a shared entity with a single component instance. My approach may not be as coherent.

_________

> Having an entity with multiple components is easy enough AIUI -- put a slice or whatever on the component struct right?

I'm not sure I follow

_____________

It's been about half a year since I looked at any of this code, so actual implementation details would require me to give this some more serious  examination of the codebase. It sounds like you're way ahead of me!

The current implementation is the original, and not very well optimized. If you feel like you have a way of making _any_ of this better, please feel free to open an MR. I am also open to further discussion. If you'd like to open another issue as a feature proposal, we can iron out any of our concerns there.
